File: app/src/main/java/net/squanchy/speaker/SpeakerDetailsActivity.java
package net.squanchy.speaker;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;

import net.squanchy.R;
import net.squanchy.navigation.Navigator;
import net.squanchy.speaker.domain.view.Speaker;
import net.squanchy.speaker.widget.SpeakerDetailsLayout;
import net.squanchy.support.config.DialogLayoutParameters;
import net.squanchy.support.lang.Optional;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;

public class SpeakerDetailsActivity extends AppCompatActivity {

    private static final String EXTRA_SPEAKER_ID = SpeakerDetailsActivity.class.getCanonicalName() + ".speaker_id";

    private CompositeDisposable subscriptions = new CompositeDisposable();
    private SpeakerDetailsService service;
    private Navigator navigator;

    private SpeakerDetailsLayout speakerDetailsLayout;

    private Optional<Speaker> speaker = Optional.absent();

    public static Intent createIntent(Context context, String speakerId) {
        Intent intent = new Intent(context, SpeakerDetailsActivity.class);
        intent.putExtra(EXTRA_SPEAKER_ID, speakerId);
        return intent;
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_speaker_details);

        DialogLayoutParameters.fullHeight(this)
                .applyTo(getWindow());

        SpeakerDetailsComponent component = SpeakerDetailsInjectorKt.speakerDetailsComponent(this);
        service = component.service();
        navigator = component.navigator();

        Toolbar toolbar = findViewById(R.id.toolbar);
        setupToolbar(toolbar);

        speakerDetailsLayout = findViewById(R.id.speaker_details_root);
    }

    private void setupToolbar(Toolbar toolbar) {
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayShowTitleEnabled(false);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setHomeAsUpIndicator(R.drawable.ic_close);
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);

        subscriptions.dispose();
        observeSpeakerFrom(intent);
    }

    @Override
    protected void onStart() {
        super.onStart();

        Intent intent = getIntent();
        observeSpeakerFrom(intent);
    }

    private void observeSpeakerFrom(Intent intent) {
        String speakerId = intent.getStringExtra(EXTRA_SPEAKER_ID);

        subscriptions = new CompositeDisposable();
        subscriptions.add(service.speaker(speakerId)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::onSpeakerRetrieved));
    }

    private void onSpeakerRetrieved(Speaker speaker) {
        speakerDetailsLayout.updateWith(speaker);

        this.speaker = Optional.of(speaker);
        supportInvalidateOptionsMenu();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.speaker_details, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        MenuItem twitterItem = menu.findItem(R.id.action_speaker_twitter);
        MenuItem websiteItem = menu.findItem(R.id.action_speaker_website);

        if (speaker.isPresent()) {
            Speaker speaker = this.speaker.get();

            twitterItem.setVisible(speaker.getTwitterUsername().isPresent());
            websiteItem.setVisible(speaker.getPersonalUrl().isPresent());
        } else {
            twitterItem.setVisible(false);
            websiteItem.setVisible(false);
        }

        return super.onPrepareOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int itemId = item.getItemId();
        if (itemId == android.R.id.home) {
            finish();
            return true;
        } else if (itemId == R.id.action_speaker_twitter) {
            navigate().toTwitterProfile(speaker.get().getTwitterUsername().get());
            return true;
        } else if (itemId == R.id.action_speaker_website) {
            navigate().toExternalUrl(speaker.get().getPersonalUrl().get());
            return true;
        } else {
            return super.onOptionsItemSelected(item);
        }
    }

    private Navigator navigate() {
        return navigator;
    }

    @Override
    protected void onStop() {
        super.onStop();
        subscriptions.dispose();
    }
}

File: app/src/main/java/net/squanchy/speaker/SpeakerDetailsActivity.kt
package net.squanchy.speaker

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.rxkotlin.plusAssign
import io.reactivex.rxkotlin.subscribeBy
import kotlinx.android.synthetic.main.activity_speaker_details.*
import net.squanchy.R
import net.squanchy.navigation.Navigator
import net.squanchy.speaker.domain.view.Speaker
import net.squanchy.support.config.DialogLayoutParameters
import net.squanchy.support.lang.getOrThrow
import timber.log.Timber

class SpeakerDetailsActivity : AppCompatActivity() {

    private val subscriptions = CompositeDisposable()

    private lateinit var service: SpeakerDetailsService
    private lateinit var navigator: Navigator

    private var speaker: Speaker? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_speaker_details)

        DialogLayoutParameters.fullHeight(this)
            .applyTo(window)

        with(speakerDetailsComponent(this)) {
            service = service()
            navigator = navigator()
        }

        setupToolbar(toolbar)
    }

    private fun setupToolbar(toolbar: Toolbar) {
        setSupportActionBar(toolbar)
        supportActionBar?.apply {
            setDisplayShowTitleEnabled(false)
            setDisplayHomeAsUpEnabled(true)
            setHomeAsUpIndicator(R.drawable.ic_close)
        }
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)

        subscriptions.clear()
        observeSpeakerFrom(intent)
    }

    override fun onStart() {
        super.onStart()

        observeSpeakerFrom(intent)
    }

    private fun observeSpeakerFrom(intent: Intent) {
        val speakerId = intent.getStringExtra(EXTRA_SPEAKER_ID)

        subscriptions +=
            service.speaker(speakerId)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeBy(
                    onNext = ::onSpeakerRetrieved,
                    onError = { e -> Timber.e(e, "Error retrieving the speaker") }
                )
    }

    private fun onSpeakerRetrieved(speaker: Speaker) {
        speakerDetailsLayout.updateWith(speaker)

        this.speaker = speaker
        invalidateOptionsMenu()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.speaker_details, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        val twitterItem = menu.findItem(R.id.action_speaker_twitter)
        val websiteItem = menu.findItem(R.id.action_speaker_website)

        twitterItem.isVisible = speaker?.twitterUsername?.isDefined() ?: false
        websiteItem.isVisible = speaker?.personalUrl?.isDefined() ?: false

        return super.onPrepareOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val itemId = item.itemId
        when (itemId) {
            android.R.id.home -> {
                finish()
                return true
            }
            R.id.action_speaker_twitter -> {
                navigator.toTwitterProfile(speaker!!.twitterUsername.getOrThrow())
                return true
            }
            R.id.action_speaker_website -> {
                navigator.toExternalUrl(speaker!!.personalUrl.getOrThrow())
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    override fun onStop() {
        super.onStop()
        subscriptions.dispose()
    }

    companion object {

        private val EXTRA_SPEAKER_ID = "${SpeakerDetailsActivity::class.java.name}.speaker_id"

        fun createIntent(context: Context, speakerId: String) =
            Intent(context, SpeakerDetailsActivity::class.java).apply {
                putExtra(EXTRA_SPEAKER_ID, speakerId)
            }
    }
}
